specs/semantics/hs/test/Control/State/Transition/Examples/RegistryModel.hs

```diff
@@ -0,0 +1,438 @@
+
+    perform
+      :: Map AThreadId ThreadId
+      -> RegCmd
+      -> m (Map AThreadId ThreadId)
+    perform st (Spawn atid) = do
+      tid <- evalIO $ forkIO (threadDelay 10000000)
+      return $! Map.insert atid tid st
+    perform st (Register name atid) = do
+      let tid = fromJust $ Map.lookup atid st
+      evalIO $ register name tid
+      return $! st
+    perform st (Unregister name) = do
+      evalIO $ unregister name
+      return st
+    perform st (WhereIs name (Just atid)) = do
+      mtid <- evalIO $ whereis name
+      let expectedTid = Map.lookup atid st
+      mtid === expectedTid
+      return st
+
+--------------------------------------------------------------------------------
+-- Let's try to hookup the system under test using state machines!
+--------------------------------------------------------------------------------
+
+prop_RegistryImplementsModel :: Property
+prop_RegistryImplementsModel = withTests 1000 $ property $ do
+  tidMapRef <- evalIO $ newIORef []
+  env <- forAll $ initEnvGen @REGISTRY
+  actions <- forAll $
+    Gen.sequential
+      (Range.linear 1 100)
+      initialState
+      [registryCmd tidMapRef env]
+  _ <- evalIO cleanUp
+  executeSequential initialState actions
+
+newtype AbstractState (v :: * -> *) = AbstractState (State REGISTRY)
+  deriving (Show)
+
+initialState :: AbstractState v
+initialState = AbstractState ([], [])
+
+data RegCmdW (v :: * -> *) = RegCmdW RegCmd deriving (Eq, Show)
+
+instance HTraversable RegCmdW where
+  htraverse _ (RegCmdW cmd) = pure (RegCmdW cmd)
+
+data RegOut
+  = Spawned ThreadId
+  | Registered
+  | Unregistered
+  | Found ThreadId
+  | Failed ErrorCall
+  | NotFound String
+  deriving (Eq, Show)
+
+isFailure :: RegOut -> Bool
+isFailure (Failed _) = True
+isFailure _ = False
+
+wasFound :: RegOut -> Bool
+wasFound (Found _) = True
+wasFound _ = False
+
+registryCmd
+  :: forall m
+   . (MonadTest m, MonadIO m)
+  => IORef (Map AThreadId ThreadId)
+  -> Environment REGISTRY
+  -> Command Gen m AbstractState
+registryCmd tidMapRef env =
+  Command
+    gen
+    execute
+    callbacks
+  where
+    -- If we're using our STS framework, we should always return a generator. I
+    -- don't know whether there are cases in which we can decide that we won't
+    -- be able to generate an action based on the initial state.
+    gen :: AbstractState v -> Maybe (Gen (RegCmdW v))
+    gen (AbstractState st) = Just $ fmap RegCmdW $ sigGen @REGISTRY env st
+
+    execute :: RegCmdW v -> m RegOut
+    execute (RegCmdW (Spawn atid)) = do
+      etid <- evalIO $ try $ forkIO (threadDelay 10000000)
+      case etid of
+        Left e -> pure $! Failed e
+        Right tid -> do
+          evalIO $ modifyIORef tidMapRef $ Map.insert atid tid
+          return $! Spawned tid
+    execute (RegCmdW (Register name atid)) = do
+      -- We need the abstract-thread-id to thread-id mapping! So I guess we
+      -- have to resort to IORefs :/
```

Comment:
This is where you should be using the `concrete` stuff, but the problem is that you're using `RegCmd` which isn't parameterised by `v`. I'm not sure how this would work here, because you'd have to make the `REGISTRY` STS stuff aware of this parameter...

Reply:
Yes, indeed. So I don't know whether embedding STS's in SM's is the right thing to do ...

Reply:
>  you'd have to make the REGISTRY STS stuff aware of this parameter...

And this is what you end up doing in your branch https://github.com/input-output-hk/cardano-ledger/tree/ruhatch/hedgehog-model  :)